images/shared-images-controller/main.go

```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list releases from kubernetes/kubernetes")
+		return err
+	}
+	minors := querier.LastThreeMinor(uint(1), releases)
+	for _, minor := range minors {
+		r := querier.ParseRelease(minor)
+		releaseVer := fmt.Sprintf("%s.%s", r.Major, r.Minor)
+		versions = append(versions, releaseVer)
+	}
+
+	log.Infoln("Last three minor releases", versions)
+
+	var imageNames []string
+	kubevirtci_repo := "quay.io/kubevirtci/"
+
+	imageList, err := images.List(ctx, nil)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list images")
+		return err
+	}
+	for _, i := range imageList {
+		for _, imageName := range i.RepoTags {
+			imageNames = append(imageNames, imageName)
+		}
+	}
+
+	for _, version := range versions {
+		log.Infoln("Kubevirt Provider version: ", version)
+		name := fmt.Sprintf("%sk8s-%s:%s", kubevirtci_repo, version, tag)
+		if checkImagePresent(imageNames, name) {
+			log.Infoln("Image already present:", name)
+		} else {
+			log.Infoln("Pulling image: ", name)
+			_, err := images.Pull(ctx, name, nil)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to pull image:", name)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func cleanOldImages(ctx context.Context, tag string) error {
+	var deleteImages []string
+	imageList, err := images.List(ctx, nil)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list images")
+		return err
+	}
+	for _, i := range imageList {
+		for _, repoTag := range i.RepoTags {
+			if strings.Contains(repoTag, tag) {
+				log.Infof("%s is a required image", repoTag)
+			} else {
+				log.Infoln("Deleting image:", repoTag)
+				deleteImages = append(deleteImages, repoTag)
+			}
+		}
+	}
+	log.Infoln("images to delete:", deleteImages)
+	if len(deleteImages) > 0 {
+		_, remove_err := images.Remove(ctx, deleteImages, nil)
+		if remove_err != nil {
+			log.WithError(err).Errorf("Failed to delete old images")
+			return err
+		}
+	}
+	return nil
+}
+
+func main() {
+
+	var period time.Duration = 6
+
+	socket := "unix:/run/podman/podman.sock"
+
+	if os.Getenv("XDG_RUNTIME_DIR") != "" {
+		sock_dir := os.Getenv("XDG_RUNTIME_DIR")
+		socket = "unix:" + sock_dir + "/podman/podman.sock"
+	}
+	connText, err := bindings.NewConnection(context.Background(), socket)
+	if err != nil {
+		log.WithError(err).Errorf("Could not connect to podman socket %d", socket)
+		os.Exit(1)
+	}
+	for {
+		tag, err := getLatestTag()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get latest kubevirtci tag")
+		}
+		err = pullRequiredImages(connText, tag)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to fetch image names")
+		}
+		err = cleanOldImages(connText, tag)
+		if err != nil {
+			log.WithError(err).Errorf("Failure occured when deleting old images")
+		}
+		log.Infof("Waiting for %d hours before checking again", period)
+		time.Sleep(period * time.Hour)
```

Comment:
When do we kill the job?

Reply:
Ok, I got confused by the addition of pre/post submit jobs. Do you want to use them for testing? If so would it make sense to implement an option to indicate you want one run and then exit?

Reply:
the pre/post submit jobs are just used for building / publishing the image that will be used by the daemonset.

Reply:
My bad. I see it now. Should have looked before.